src/Core/BinReader.fs

```diff
@@ -54,14 +54,41 @@ type BinReader (bytes: byte []) =
   /// Peek a character array of size n at the given offset.
   member __.PeekChars (n: int, o: int) = __.PeekBytes (n, o) |> Array.map char
 
-  /// Peek a number (64-bit integer) using the ULEB128 method.
-  member __.PeekULEB128 (o: int) =
-    let rec readLoop n cnt offset  =
-      let b = __.PeekByte (offset)
-      let acc = (uint64 (b &&& 0x7Fuy) <<< (cnt * 7)) ||| n
-      if b &&& 0x80uy <> 0uy then readLoop acc (cnt + 1) (offset + 1)
-      else acc, cnt + 1
-    readLoop 0UL 0 o
+  /// Peek a LEB128-encoded integer into uint64. This function returns a tuple of
+  /// (the decoded uint64, and the count of how many bytes were peeked).
+  member __.PeekUInt64LEB128 (o: int) =
+    let len = if __.Bytes.Length > LEB128.max64LEB128Length
+              then LEB128.max64LEB128Length
+              else __.Bytes.Length
```

Comment:
This function tries to peek a value from the given offset, and you should not peek the whole file contents (i.e., the `len` should not be `__.Bytes.length`.

Therefore, you cannot simply check if `__.Bytes.Length > LEB128.max64LEB128Length`. Instead, you should check if you have enough contents after the given offset up to the end of the file.

Although this seems "straightforward", there is a serious problem here, though. Since we can only expect the maximum possible length to fetch, this function can fail badly when we try to peek a LEB128 value located at the end of the file. For example, suppose we have a 3-byte LEB128 value at the end of a file. In this case, this function will try to `__.PeekBytes` from an offset `length_of_file - 1 - 3`. The function will then try to optimistically fetch a byte array of the max length (`LEB128.max32LEB128Length`) from the offset, which will raise an exception as we will access beyond the scope of the `__.Bytes` array.

So, it is not a good idea to directly use the LEB128 functions in BinReader. Instead, we should keep the `readLoop` implementation as before, but move the `readLoop` out of the member function, so that we can use the same function for all the other functions. We then use the [`extendSign` function](https://github.com/B2R2-org/B2R2/blob/master/src/Core/LEB128.fs#L47) of the `LEB128` module for peeking signed ULEB128 values.

Hopefully this makes sense to you.

Reply:
(1) I'm a bit confused by the last point (readLoop Implementation), please can you explain what do you mean with:
 > move the readLoop out of the member function, so that we can use the same function for all the other functions.

(2) I have a little suggestion so I hope you like it: I'm thinking of adding a generic function to the LEB128 module that calculates the count of how many bytes to peek/read, it can be very useful for bounds checking.

Reply:
(1) I meant you want to refactor the code.  The `readLoop` pattern will occur multiple times in the other __.PeekXXXLEB128 member functions too. So you want to make the `readLoop` to be defined as a separate (private) member function.

(2) If I understood correctly, the count function you are talking about should be similar to the `readLoop` function in the original implementation. Is there a particular reason to place the function in the LEB128 module other than just put it inside `BinReader` as a private member as I described in (1)? I don't know if it is useful to export the count function because it is not likely to be used by others (other than the `BinReader` module.) Am I missing something here?

Reply:
(1) Never mind about my suggestion, I realized it was useless.
(2) Talking about readLoop function, If I understood correctly that means the resulting is just a re-implementation of LEB128 decoding functions, if so what it's the point of creating the LEB128 module from the beginning, Am I missing something ?
I really appreciate your patience, thanks for your time!

Reply:
You are right, but we may want to decode a LEB128 value not only from a file data, but also from some other sources. For example, we can put the LEB128 decoding feature in our REPL.

Nevertheless, it is unfortunate that we have to implement the similar logic twice, but the function itself is so simple and not a big deal in my opinion.

Reply:
This will be problematic, since the length of the bytes sequence is unknown this will make overflow detection much harder, for example consider the following byte arrays (Taken from `LEB128` tests): `[| 0xffuy; |]` and `[| 0x80uy; 0x80uy; |]`,
if we tried to decode them into a uint64 value we'll not be able to detect the overflow and as a result we'll get an invalid values, so the old `PeekULEB128` implementation has a bug, Please let me know if I'm wrong.

Reply:
BinReader module is designed for reading a binary file, and the PeekULEB128 is for reading a LEB128 encoded value from a binary file. If there is a certain field in a binary file that is known to be encoded in LEB128, we can safely assume that the value is valid. Otherwise, the file is broken.

Therefore PeekULEB128 doesn't need to do the overflow check, unlike the functions in the LEB128 module.